### ISSUE TITLE
Binary AND returns incorrect value

### ISSUE THREAD

**Issue by MikeM-MSFT** [PowerShell/PowerShell#4031]
Steps to reproduce
------------------

[Convert]::ToString(0xFFFFFFFFFFFF -band 0xFF0000000000,2)
[Convert]::ToString(0xFFFFFFFFFFFF -band 0x00FF00000000,2)
[Convert]::ToString(0xFFFFFFFFFFFF -band 0x0000FF000000,2)
[Convert]::ToString(0xFFFFFFFFFFFF -band 0x000000FF0000,2)
[Convert]::ToString(0xFFFFFFFFFFFF -band 0x00000000FF00,2)
[Convert]::ToString(0xFFFFFFFFFFFF -band 0x0000000000FF,2)

Expected behavior
-----------------
111111110000000000000000000000000000000000000000
1111111100000000000000000000000000000000
11111111000000000000000000000000
111111110000000000000000
1111111100000000
11111111

Actual behavior
---------------

111111110000000000000000000000000000000000000000
1111111100000000000000000000000000000000
111111111111111111111111000000000000000000000000
111111110000000000000000
1111111100000000
11111111

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.15063.413
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.15063.413
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

**Comment by mklement0** [comment]
Poor `-band` is just the messenger here: **the problem is more fundamental, namely PowerShell's lack of _unsigned_ integer types**, which leads to unexpected behavior with hex. constants:

In short:

* Use `[Convert]::ToString(0xFFFFFFFFFFFF -band 0x0000FF000000L,2)` (note the `L` suffix) to _work around_ the problem.

* Even with the workaround, hex. constants only work as intended up to `[int64]::MaxValue` (`0x7fffffffffffffff`), not `[uint64]::MaxValue` (`0xffffffffffffffff`)

* Contravening the fundamental user expectation that a hex. literal is always a *positive* - unsigned - number is problematic - see #3313

---

As for the specifics of the symptom at hand:

```powershell
# The smallest *signed* integer type that can fit the bit pattern is chosen, which 
# is Int32 in this case.
# The problem is that this can - unexpectedly result in a *negative* number:
> [Convert]::ToString(0x0000FF000000,2);
 (0x0000FF000000).GetType().Name; 0x0000FF000000;
11111111000000000000000000000000  # so far, so good - the *32-bit* pattern is correct
Int32
-16777216  # !! Result is negative, because [System.Int32] was chosen as the data type.
```

Similarly, `0xFF0000000000` becomes _signed_ type `[System.Int64]` (`long`).
Therefore, in expression `0xFFFFFFFFFFFF -band 0x0000FF000000`, the RHS, an `Int32` by itself, is implicitly coerced to the type of the larger LHS, `Int64`, which is when the problem surfaces:

```powershell
> [Int64] 0x0000FF000000
-16777216  # Result is the same - negative - value as before, only as Int64

# That Int64 representation does NOT have the expected bit pattern, however, 
# due to being a signed type:
> [Convert]::ToString([Int64] 0x0000FF000000, 2)
1111111111111111111111111111111111111111000000000000000000000000

# Here's the expected pattern, stemming from forcing the literal to an long (L), aka Int64.
> [Convert]::ToString(0x0000FF000000L, 2)
11111111000000000000000000000000
```



P.S.: Can you please edit your post to use fenced code / output blocks for readability?

**Comment by iSazonov** [comment]
User can catch the issue with using C# or C++ binary flag constants. 😕

**Comment by vexx32** [comment]
This is likely to have been inadvertently fixed with either #7813 or #7993 where we added proper support for unsigned hex and binary literals.

These changes are not being backported to PS 5.1 -- almost nothing is able to be backported to the Windows PowerShell edition. I'm closing this as resolved for now, thanks for following up and confirming it's no longer an issue in 7.2+ 🙂
